Common/src/main/java/com/blamejared/crafttweaker/api/util/ParseUtil.java

```diff
@@ -1,54 +1,71 @@
 package com.blamejared.crafttweaker.api.util;
 
+import com.blamejared.crafttweaker.natives.resource.ExpandResourceLocation;
+import net.minecraft.resources.ResourceLocation;
 import org.openzen.zencode.shared.CodePosition;
+import org.openzen.zencode.shared.LiteralSourceFile;
 import org.openzen.zenscript.lexer.ParseException;
 import org.openzen.zenscript.lexer.ZSTokenParser;
 import org.openzen.zenscript.lexer.ZSTokenType;
+import org.openzen.zenscript.parser.expression.ParsedCallArguments;
 import org.openzen.zenscript.parser.expression.ParsedExpression;
 import org.openzen.zenscript.parser.expression.ParsedExpressionMember;
+import org.openzen.zenscript.parser.expression.ParsedExpressionString;
 import org.openzen.zenscript.parser.expression.ParsedExpressionVariable;
+import org.openzen.zenscript.parser.expression.ParsedNewExpression;
 import org.openzen.zenscript.parser.type.IParsedType;
-import org.openzen.zenscript.parser.type.ParsedNamedType;
+import org.openzen.zenscript.parser.type.ParsedTypeBasic;
 
-import java.util.Arrays;
+import java.io.IOException;
+import java.util.ArrayList;
 import java.util.List;
-import java.util.stream.Collectors;
+import java.util.regex.Pattern;
 
 public final class ParseUtil {
-
+    
+    private static final Pattern typeArgumentPattern = Pattern.compile("(.*)<(.*)>");
+    
     private ParseUtil() {
+    
     }
-
+    
     public static ParsedExpression explode(CodePosition position, String name) {
-
+        
         String[] strExpressions = name.split("\\.");
         ParsedExpression expression = new ParsedExpressionVariable(position, strExpressions[0], null);
-        if (strExpressions.length > 1) {
-            for (int i = 1; i < strExpressions.length; i++) {
+        if(strExpressions.length > 1) {
+            for(int i = 1; i < strExpressions.length; i++) {
                 expression = new ParsedExpressionMember(position, expression, strExpressions[i], List.of());
             }
         }
-
+        
         return expression;
     }
-
-    public static IParsedType readParsedType(String name, CodePosition position) {
-
-        final List<ParsedNamedType.ParsedNamePart> collect = Arrays.stream(name.split("[.]"))
-                .map(s -> new ParsedNamedType.ParsedNamePart(s, null))
-                .collect(Collectors.toList());
-
-        return new ParsedNamedType(position, collect);
+    
+    public static IParsedType readParsedType(String name, CodePosition position) throws ParseException {
+        
+        try {
+            return IParsedType.tryParse(ZSTokenParser.create(new LiteralSourceFile("type reading: " + name, name), null));
+        } catch(IOException e) {
+            e.printStackTrace();
```

Comment:
That exception should actually never be hit, it can only occur when something goes wrong reading from the file, but the file in this case is an in memory string, so I'm not sure how it could ever fail.